### ISSUE TITLE
Request Feedback on VSCode Extension (Milestone #1)

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#1374]
Hello everyone,

We are pleased to announce the release of milestone 1 of the VSCode extension.

We would like your feedback to make sure the extension is robust and usable.

We ask that you install the extension and then experiment with the following features (other features are available, but we will request feedback for those at a later time).:

- [ ] Hover over any expression to see its type and effect.
- [ ] Hover over any local variable or formal parameter to see its type.
- [ ] Hover over any function to see its type signature and documentation. 
- [ ] Jump to the definition of any function.
- [ ] Jump to the definition of any local variable or formal parameter.
- [ ] Jump to the definition of any enum case.
- [ ] Find all references to a function.
- [ ] Find all references to a local variable or formal parameter.
- [ ] Find all references to an enum case.
- [ ] Does the editor stay in synch when editing, creating, or deleting files?

Please select all that you tested and that worked as expected.

Now, we would like to ask you a few more questions:

- Did you have any issues with the installation?

- Did you notice any problems with the syntax highlighting? 

- Did the diagnostics (error messages) show up/disappear when expected?

- How would you rate the _startup time_ of the plugin?

- How would you rate the _responsiveness_ of the plugin?

- Did you encounter any errors with the plugin?

- Did you find the UX intuitive?

- If you read the README.md, do you have any suggestions for improvements?

- Would you prefer that instead of an internally globally stored flix.jar, we always use and download flix.jar to the current project directory?

- How often should we run the compiler? Currently we run it 500ms after the user has stopped typing. We can run it more often at the cost of more CPU usage.

- We believe that we can implement the following features:
  * Selection Range (makes it more easy to select/navigate expressions).
  * Semantic Token (improve syntax highlighting slightly).
  * Rename support (local variables, functions, enum tags, - not yet types).
  * Document Highlights Request (e.g. when the cursor is on a local variable all references are highlighted).
  * Document Symbols (provides a list of functions defined in the current file. Can search/jump based on name).
  * Proper output from runMain, runBenchmarks, and runTests.
  * If you had to pick one or two, which would you pick for milestone 2?

(Obviously everyone wants auto-complete, but we cannot yet implement that. Its on the roadmap.)

- Do you have any other comments?

**Comment by magnus-madsen** [comment]
@JonathanStarup @esbenbjerre @stephentetley @mlutze If you have the time, I would very much like to hear from each of you :)

**Comment by esbenbjerre** [comment]
I can't figure out how to run Flix code inside VS Code. IntelliJ seems to magically know how to build everything correctly and run main/src/ca/uwaterloo/flix/Main.scala, but I can't get VS Code to do the same thing. Would anybody mind sharing a launch/tasks.json file that does this?

**Comment by magnus-madsen** [comment]
Did you install the VSCode plugin from the market? It should automatically download Flix.

**Comment by magnus-madsen** [comment]
See also https://github.com/flix/flix.dev/blob/master/src/gif/vscode.png

**Comment by esbenbjerre** [comment]
TL;DR it works now.

I had to do a complete wipe and reinstall of VS Code to get CodeLens working.

**Comment by JonathanStarup** [comment]
All the noted points works well for me although I didn't spend enough time with it to judge load times / responsiveness / sync issues. I had some issues with installation (but they were easily fixed). The error:
```
Flix LSP responded with an error. 
P
Flix LSP responded with an error. 
P
Flix LSP responded with an error. 
icked up _JAVA_OPTIONS: -Xmx512M

Flix LSP responded with an error. 
icked up _JAVA_OPTIONS: -Xmx512M

Could not read file (file:///Users/auXXXXXX/Desktop/test.flix) in queue. 
Error: TypeError [ERR_INVALID_FILE_URL_PATH]: File URL path must be absolute
Could not read file (file:///Users/auXXXXXX/Desktop/test.flix) in queue. 
Error: TypeError [ERR_INVALID_FILE_URL_PATH]: File URL path must be absolute
Flix Extension 0.8.0 ready! Running Flix 0.13-rev0
(c:\Users\auXXXXXX\AppData\Roaming\Code\User\globalStorage\flix.flix\flix.jar)
```
When I removed the _JAVA_OPTIONS from my path and loaded a file that was not on a network location it worked fine. I don't know if the remote file issue is just a vscode thing.

**Comment by magnus-madsen** [comment]
We are also looking for Flix projects to try out the VSCode plugin on. Currently we have:

- https://github.com/flix/actors

But if anyone has self-contained projects that we can use to test with, we would like to know :)

**Comment by stephentetley** [comment]
![logo](https://user-images.githubusercontent.com/12343071/95682550-b064b500-0bdd-11eb-9da1-d9d41aca740c.png)

The extension looks great, though it's missing an icon so I've cropped the Flix icon and added a transparent background.

The path to the icon should go in `package.json` at the same level as "name", "license" etc. The key is "icon".

**Comment by esbenbjerre** [comment]
-[x] Hover over any expression to see its type and effect.
-[x]  Hover over any local variable or formal parameter to see its type.
-[x] Hover over any function to see its type signature and documentation.
-[x]  Jump to the definition of any function.
-[x]  Jump to the definition of any local variable or formal parameter.
-[x]  Jump to the definition of any enum case.
-[x]  Find all references to a function.
-[x]  Find all references to a local variable or formal parameter.
-[x]  Find all references to an enum case.
-[x]  Does the editor stay in synch when editing, creating, or deleting files?

**Did you have any issues with the installation?**
Yes, at first the extension crashed at the "Starting up" message. However, it seemed to be caused by my own broken VS Code installation.

**Did you notice any problems with the syntax highlighting?**
The list concat operator `:::` is highlighted like cons, but I see that has been filed as a bug in the repo already.

**Did the diagnostics (error messages) show up/disappear when expected?**
Yes.

**How would you rate the startup time of the plugin?**
Fair. The plugin takes around 5 seconds to initialize on my MacBook Pro from 2013.

**How would you rate the responsiveness of the plugin?**
Fair. After introducing a syntax error it takes the plugin half a second to up

**Did you encounter any errors with the plugin?**
Yes. A file located directly in the desktop folder on macOS will not run. It crashes with
```
Flix LSP responded with an error. 
java.nio.file.FileSystemException: /./target: Read-only file system

Flix LSP responded with an error. 
java.nio.file.FileSystemException: /./target: Read-only file system

Flix LSP responded with an error. 
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:100)

Flix LSP responded with an error. 
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
	at java.base/sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:389)
	at java.base/java.nio.file.Files.createDirectory(Files.java:694)
	at java.base/java.nio.file.Files.createAndCheckIsDirectory(Files.java:801)
	at java.base/java.nio.file.Files.createDirectories(Files.java:787)
	at ca.uwaterloo.flix.language.phase.jvm.JvmOps$.writeClass(JvmOps.scala:1185)
	at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.$anonfun$run$6(JvmBackend.scala:184)
	at ca.uwaterloo.flix.language.phase.jvm.JvmBackend$.$anonfun$run$6$adapted(JvmBackend.scala:183)
```

**Did you find the UX intuitive?**
Yes.

**If you read the README.md, do you have any suggestions for improvements?**
No, seems clear and concise.

**Would you prefer that instead of an internally globally stored flix.jar, we always use and download flix.jar to the current project directory?**
No, I think that is a non-issue. Users wishing to use a specific flix.jar can just place it in the root project.

**How often should we run the compiler? Currently we run it 500ms after the user has stopped typing. We can run it more often at the cost of more CPU usage.**
From my own experience I think it would be possible to shave it down to 250ms without sacrificing too much CPU usage.

For milestone 2 I would prefer to see Rename support and Document Highlights Request.

**Comment by magnus-madsen** [comment]
Thanks for the feedback. The plugin seems to work reasonably well now.

We have plans to work on auto-complete over the summer.
